### DISCUSSION TITLE
GLaBIOS - Cassette BASIC doesn't boot and BASIC.COM doesn't run

### DISCUSSION THREAD

**Discussion by Torinde** [86Box/86Box#4772] [Q&A] [answered]
1. If I put the Floppy controller to "Internal", but don't give an image file - it doesn't boot into BASIC (maybe the "machine" doesn't have ROM BIOS associated with it), but just restarts after the "Disk Boot Fail" monster message.
2. If I give it a PC DOS 1.00 image it boots to DOS prompt and works. But if I run BASIC or BASICA - it starts to beep-beep-beep without stopping.
3. It seems the GLaBIOS machine is XT. Can you please add PC (5150) as well? With 16KB-64KB RAM and Cassette media.

https://github.com/640-KB/GLaBIOS/discussions/50#discussioncomment-10488935, #4548

**Comment by jriwanek** [comment, marked as answer]
GlaBIOS is a configurable BIOS choice on the pc81 and pc82 in latest

**Reply by Torinde** [reply]
Thanks, all three items above work OK now!

Could you please enable the hybrid 16/32KB configuration as well? It's described in the linked discussions - what is needed is a second RAM dropdown for IBM PC, because there you can have the **jumpers** set to one value (e.g. 16KB), while _actually_ the **sockets** are filled with different amount of RAM (e.g. 32KB or more). That allows booting operating systems with less than 32KB RAM.

**Reply by jriwanek** [reply]
Seems silly, and needless, other than misreporting the RAM, it's not actually booting systems with less than 32K of RAM.

**Reply by Torinde** [reply]
Per my understanding - the jumper setting is what informs the BIOS (and from there DOS) on the amount of memory (e.g. 16KB), thus operating system (and applications relying on it for memory allocation/check) see only 16KB. What prevents us from doing that in 86Box is that the bootsector is written just below 32KB and that works on a real PC (with the second socket populated), but not on 86Box.

**Reply by jriwanek** [reply]
You can pick 16kb on the PC81 system, I fail to see any benefit to what you're requesting, unless you can make a better case than "I want it", then our answer is no.

**Reply by jriwanek** [reply]
Like, at bare minimum actually list some applications/operating systems which need this, which also need the option rom support from a pc82 BIOS, not just a vague "operating systems/applications".

**Reply by Torinde** [reply]
When you pick 16KB on PC81 and put a floppy image - you get a blinking cursor (while a real PC gives `PARITY CHECK 1` or DOS booting - depending on absence/presence of RAM chip in the second bank).

Please see [here](https://www.minuszerodegrees.net/5150/ram/5150_not_enough_ram_for_DOS.htm) for how a real PC behaves with 16KB RAM. Its [U12 memory transceivers are permanently connected](https://www.minuszerodegrees.net/5150/motherboard/IBM%205150%20motherboard%20-%20Removing%20RAM%20chips.png) and that allows booting from 0x7C00 while jumpers are set to 16KB.

**Reply by jriwanek** [reply]
The reason you don't see parity error when doing that is because the RAM contents arn't random on 86box, this has *nothing* to do with reporting 16k when 32k is installed, 640KB basically explained this on the linked posts.

**Reply by jriwanek** [reply]
Also, AFAIK, we don't actually implement Parity to begin with, but yeah, either way, nothing to do with setting switches to a different RAM configuration than the system has.

**Reply by Torinde** [reply]
> list some applications/operating systems which need this

Isn't it the same for 32KB (and other items/features already emulated)? What program runs on 32KB, but not on 64KB, etc.
I assume you can frame the whole 86Box as a want instead of need. :)

So, the need/want is to emulate IBM PC behavior that allows booting when switches are set to 16KB. I'm trying with PC DOS so far (and the report linked as well), but I assume you can do that with CP/M, etc.

**Reply by jriwanek** [reply]
Yeah, you're not making a valid argument for your request, and moving to straw men arguments instead, 16k, 32k, 48k and 64k were standard possible configurations on the 16-64k PC 5150, the parity error (Or rather lack thereof) as specified has nothing to do with the switch settings, and you've failed to provide a use case which makes sense, beyond frankly a baffling level of pedantry (Still on things which have nothing to do with purposely setting the memory configuration jumper wrong at that), you have been told no, you have been told why. this discussion will now be closed.

**Reply by Torinde** [reply]
> you don't see parity error when doing that is because the RAM contents arn't random on 86box

OK, thanks!

640KB also said he agrees with my understanding of the jumper behavior:
> Reading [your comment here](https://github.com/joncampbell123/dosbox-x/issues/3664#issuecomment-2182430832), that's pretty much what I was saying above, so I'd say that's accurate.

So, we have both the real hardware observation and explanation of why it works. But none of the emulators do that yet.

**Reply by Torinde** [reply]
Thanks for considering it and pity it won't happen. What my personal use case is - to run PC DOS at 16KB:
![image](https://github.com/user-attachments/assets/f652001f-d5de-43f6-9be9-908ccbbac001)

And to experience the lowest possible configuration, to see what BASIC/BASICA can be made to do with so little RAM, etc.

**Reply by jriwanek** [reply]
Which, having 32k installed but only 16k reported *isnt* in fact having PC DOS running at 16k.

**Reply by Torinde** [reply]
> BIOS POST will "register" (as a limit?) only 16KB total RAM (because it's directed so by the SW1/2) ...
>
> It seems CPU and BIOS allow to use addresses beyond the POST limit, because the boot sector code at 0x7C00 gets executed.
> 
> Then, I _assume_ DOS receives the `MEMORY_SIZE`, `IO_RAM_SIZE` or similar from the BIOS and [denies loading of programs](https://stackoverflow.com/questions/3715618/how-does-dos-load-a-program-into-memory) declaring a need for more than that (for PC DOS 1.x the COMMAND.COM has the role to [load executables](https://wiki.osdev.org/MZ)).

That's what I'm trying to do/test. It seems in that situation DOS will behave _as if_ there are only 16KB RAM. You're right that it's not the same as running with _actually_ only 16KB. For that you need a non-standard BIOS booting from below 16KB and 640KB said:
> That would be interesting to try. ... I'll give it a try sometime

So, until somebody makes such BIOS - the most we can have is emulation of the U12 and 16KB/32KB hybrid setting.

**Reply by jriwanek** [reply]
[SETRAM.SCR.txt](https://github.com/user-attachments/files/18824167/SETRAM.SCR.txt)
The attached file will modify the BDA after DOS has loaded to make it report 16K, you'll have to use DEBUG.COM to generate the actual binary however. (Instructions included in the file)

**Reply by Torinde** [reply]
It seems PC DOS 1.00 DEBUG can't handle that, so I used 6.22 `DEBUG < SETRAM.DSF` (after I removed the ; comments). Result is:
![image](https://github.com/user-attachments/assets/25bef5ae-f2b6-4529-a1a1-4603999203f6)

But it doesn't seem to work under PC DOS 1.00:
![image](https://github.com/user-attachments/assets/21df62bf-2171-4e7b-8ebf-7eda77b3fb18)

Did I made an error in creating the binary? Or it needs higher version of DOS?

**Reply by jriwanek** [reply]
I may have made a screwup, looking into it.

**Reply by Torinde** [reply]
What I see [here](https://www.fysnet.net/rombios.htm):
- 0048h Interrupt Vector for INT 12h (memory size)
- 0413h usable memory size in kilobytes (interrupt 12h) 
  - (not applicable for our case) Usually set to 0280h. If less, then denotes 640k-difference=size of Extended BIOS Data Area at (9FC0:0000h)
- 0044h Interrupt Vector for INT 11h (equipment list)
- 0410h equipment list reported by interrupt 11h (see figure 1)
  - bit2&3: Amount of system board RAM: 00 - 16k, 01 - 32k, 10 - 48k, 11 - 64k

As for potential option to emulate IBM PC behavior - what about adding "18KB" RAM in the dropdown - special case, 16KB (soldered on motherboard) with one extra memory chip at the top of 32KB addresses (e.g. socket at 0x7800-0x7FFF), value to be reported/assigned via BIOS/POST/DOS to be 16KB (corresponding to the DIP switches on the motherboard). Scenario: "User didn't have enough RAM chips or some of them failed".

**Reply by Torinde** [reply]
The last comment [here](https://www.reddit.com/r/DOS/comments/10syxqg/dos_330_how_do_i_check_the_memory/?rdt=50601) is a guide to recreating the MEM command in DOS3.3 - it doesn't refer to 0413h, but to following the MCB chain, etc.

So, it seems not easy to patch-on-the-fly the memory size (all BIOS/DOS sources [are available](https://github.com/joncampbell123/dosbox-x/issues/3664#issuecomment-2182430832), but I assume replicating DOS initialization to re-arrange the memory setup will be cumbersome?)
Do you see a straightforward way to do it?

Emulating the 16K+2K behavior seems preferable, because then you rely on DOS itself to do the allocations/MCB setup, etc.
